Compilation error using Clang 16

https://www.redhat.com/en/blog/new-warnings-and-errors-clang-16
Warning `incompatible-function-pointer-types` has been upgraded to an error. As a result, the gem native extension does not build.

```
$ clang --version
Apple clang version 15.0.0 (clang-1500.3.9.4)
Target: arm64-apple-darwin23.4.0
Thread model: posix
InstalledDir: /Library/Developer/CommandLineTools/usr/bin
```

```
$ gcc --version
Apple clang version 15.0.0 (clang-1500.3.9.4)
Target: arm64-apple-darwin23.4.0
Thread model: posix
InstalledDir: /Library/Developer/CommandLineTools/usr/bin
```

```
$ ruby --version
ruby 3.2.3 (2024-01-18 revision 52bb2ac0a6) [arm64-darwin23]
```

```
gem install hive_geoip2 -v 0.1.2 -- --with-maxminddb-include=$(brew --prefix libmaxminddb)/include/ --with-maxminddb-lib=$(brew --prefix libmaxminddb)/lib
```

fails with

```
compiling hive_geoip2.c
hive_geoip2.c:102:18: warning: passing 'const uint8_t *' (aka 'const unsigned char *') to parameter of type 'const char *' converts between pointers to integer types where one is of the unique plain 'char' type and the other is not [-Wpointer-sign]
      *ret_obj = rb_enc_str_new(
                 ^~~~~~~~~~~~~~~
~/.asdf/installs/ruby/3.2.3/include/ruby-3.2.0/ruby/internal/encoding/string.h:335:24: note: expanded from macro 'rb_enc_str_new'
      rb_enc_str_new) ((str), (len), (enc)))
                       ^~~~~
hive_geoip2.c:160:22: warning: initializing 'const uint8_t *' (aka 'const unsigned char *') with an expression of type 'const char[17]' converts between pointers to integer types where one is of the unique plain 'char' type and the other is not [-Wpointer-sign]
      const uint8_t *hex = "0123456789abcdef";
                     ^     ~~~~~~~~~~~~~~~~~~
hive_geoip2.c:186:22: warning: initializing 'const uint8_t *' (aka 'const unsigned char *') with an expression of type 'const char[17]' converts between pointers to integer types where one is of the unique plain 'char' type and the other is not [-Wpointer-sign]
      const uint8_t *hex = "0123456789abcdef";
                     ^     ~~~~~~~~~~~~~~~~~~
hive_geoip2.c:213:31: warning: passing 'uint8_t[size * 2 + 1]' (aka 'unsigned char[size * 2 + 1]') to parameter of type 'const char *' converts between pointers to integer types where one is of the unique plain 'char' type and the other is not [-Wpointer-sign]
      *ret_obj = rb_cstr2inum(buf, 16);
                              ^~~
~/.asdf/installs/ruby/3.2.3/include/ruby-3.2.0/ruby/internal/intern/bignum.h:171:32: note: passing argument to parameter 'str' here
VALUE rb_cstr2inum(const char *str, int base);
                               ^
hive_geoip2.c:169:21: warning: implicit conversion loses integer precision: 'size_t' (aka 'unsigned long') to 'int' [-Wshorten-64-to-32]
      for (i = size - 1; i >= 0; i--) {
             ~ ~~~~~^~~
hive_geoip2.c:205:21: warning: implicit conversion loses integer precision: 'size_t' (aka 'unsigned long') to 'int' [-Wshorten-64-to-32]
      for (i = size - 1; i >= 0; i--) {
             ~ ~~~~~^~~
hive_geoip2.c:310:16: error: incompatible function pointer types passing 'void (VALUE)' (aka 'void (unsigned long)') to parameter of type 'VALUE (*)(VALUE)' (aka 'unsigned long (*)(unsigned long)') [-Wincompatible-function-pointer-types]
    rb_protect(guard_parse_data_list, (VALUE)&args, &exception);
               ^~~~~~~~~~~~~~~~~~~~~
~/.asdf/installs/ruby/3.2.3/include/ruby-3.2.0/ruby/internal/intern/proc.h:349:26: note: passing argument to parameter 'func' here
VALUE rb_protect(VALUE (*func)(VALUE args), VALUE args, int *state);
                         ^
6 warnings and 1 error generated.
make: *** [hive_geoip2.o] Error 1
```

However, downgrading the error to warning bypasses the issue:
```
$ gem install hive_geoip2 -v 0.1.2 -- --with-maxminddb-include=$(brew --prefix libmaxminddb)/include/ --with-maxminddb-lib=$(brew --prefix libmaxminddb)/lib --with-cflags="-Wno-error=incompatible-function-pointer-types"
Building native extensions with: '--with-maxminddb-include=/opt/homebrew/opt/libmaxminddb/include/ --with-maxminddb-lib=/opt/homebrew/opt/libmaxminddb/lib --with-cflags=-Wno-error=incompatible-function-pointer-types'
This could take a while...
Successfully installed hive_geoip2-0.1.2
1 gem installed
```

--- a/ext/hive_geoip2/hive_geoip2.c
+++ b/ext/hive_geoip2/hive_geoip2.c
@@ -8,7 +8,7 @@
 VALUE rb_mHive;
 VALUE rb_cGeoIP2;
 
-static void guard_parse_data_list(VALUE arg);
+static VALUE guard_parse_data_list(VALUE arg);
 static MMDB_entry_data_list_s * parse_data_list(
   MMDB_entry_data_list_s *data_list, VALUE *ret_obj
 );
@@ -30,10 +30,12 @@ struct args_parse_data_list {
   VALUE *ret_obj;
 };
 
-static void guard_parse_data_list(VALUE arg) {
+static VALUE guard_parse_data_list(VALUE arg) {
   struct args_parse_data_list *args = (struct args_parse_data_list *)arg;
   
   parse_data_list(args->data_list, args->ret_obj);
+
+  return RUBY_Qnil;
 }
 
 static MMDB_entry_data_list_s *

